### ISSUE TITLE
unsigned char vs float

### ISSUE THREAD

**Issue by np2001** [davisking/dlib#1611]
`
std::vector<matrix<unsigned char>> training_images;
	std::vector<unsigned long>         training_labels;
	std::vector<matrix<unsigned char>> testing_images;
	std::vector<unsigned long>         testing_labels;
	load_mnist_dataset("d:/mnist", training_images, training_labels, testing_images, testing_labels);
	
	
	std::vector<matrix<float>> training_images_f;

	training_images_f.reserve(60000);

	for (const auto& ti : training_images)
	{
		matrix<float> ti_f;
		ti_f.set_size(ti.nr(), ti.nc());

		for (int r = 0; r < ti.nr(); ++r)
		{
			for (int c = 0; c < ti.nc(); ++c)
			{
				ti_f(r, c) = ti(r, c);
			}
		}

		training_images_f.push_back(ti_f);
	}

	using net_type = loss_multiclass_log<
		fc<10,
		relu<fc<84,
		relu<fc<120,
		max_pool<2, 2, 2, 2, relu<con<16, 5, 5, 1, 1,
		max_pool<2, 2, 2, 2, relu<con<6, 5, 5, 1, 1,
		input<matrix<float>>
		>>>>>>>>>>>>;
	
	net_type net;
	dnn_trainer<net_type> trainer(net);
	trainer.set_learning_rate(0.01);
	trainer.set_min_learning_rate(0.00001);
	trainer.set_mini_batch_size(128);
	trainer.be_verbose();
	trainer.set_synchronization_file("mnist_sync", std::chrono::seconds(20));
	trainer.train(training_images_f, training_labels);

	std::vector<unsigned long> predicted_labels = net(training_images_f);
	int num_right = 0;
	int num_wrong = 0;
	// And then let's see if it classified them correctly.
	for (size_t i = 0; i < training_images.size(); ++i)
	{
		if (predicted_labels[i] == training_labels[i])
			++num_right;
		else
			++num_wrong;

	}
	cout << "training num_right: " << num_right << endl;
	cout << "training num_wrong: " << num_wrong << endl;
	cout << "training accuracy:  " << num_right / (double)(num_right + num_wrong) << endl;
	`

## Expected Behavior
This code fragment is an essential of an Deep Learning Introduction Part 1 example, except for it uses training_images_f vector of float matrices as an input. Input values are in range of [0.0;255.0] and copied from original training data. Expected to get similar training results as in original example.

## Current Behavior
Loss value stops decreasing after just a few epochs. Training accuracy is below 12%.

## Steps to Reproduce
Use float matrix with values in range [0.0;255.0] as an input data. If data values are converted to range of [0.0;1.0] traiing results become similar to original example. Are float and char values treated differently or some internal conversion is performed?

* **Version**: 19.16.99 built with Cuda 10.0 support
* **Where did you get dlib**: https://github.com/davisking/dlib/
* **Platform**: Windows 10 64 bit
* **Compiler**: Visual Studio 2017(15.9.3)

**Comment by davisking** [comment]
Yes, this is what it does.  This is discussed in the documentation: http://dlib.net/dlib/dnn/input_abstract.h.html#input

**Comment by davisking** [comment]
Also, http://dlib.net/faq.html#Whereisthedocumentationforobjectfunction

**Comment by np2001** [comment]
Thanks, your documentation is pretty complete, but it's style is not yet too familiar for me. In this particular case, I was confused because "unsigned char" looks like an standard numeric type, which values are should be used "as is, without conversion". You have, for example, rgb_pixel type, which is much more understandable to be converted. Why not to define a kind of "gray_pixel" type to use in such cases?

**Comment by np2001** [comment]
It will be differ just as `rgb_pixel` differs from `unsigned char[3]`. It maybe the same at low-level, but may represent other kind of data. Since your library is not intended to be used only for images, `unsigned chars` may also represent text chars or even some raw numbers, which are not have to be divided by 256 before usage. However, it maybe just my way of thinking, or I just missing some common convention :)

**Comment by davisking** [comment]
I think the current setup is what most people expect.

But you can make a new input layer easily that behaves any way you like.

**Comment by np2001** [comment]
Well, thanks anyway. You're making a great job.
